### ISSUE TITLE
Rum support for Stylefy

### ISSUE THREAD

**Issue by Azzurite** [Jarzka/stylefy#56]
I see that Stylefy currently only supports Reagent. It would be very cool if it could support Rum as well.

(I wanted server-side-rendering and it seems with Reagent, that's not easy, while with Rum, it's trivial)

---

**How?**

Upon very first glance, it seems like it *could* be very simple. *(With a lot of caveats: I've read the Stylefy source, but don't have a full understanding; what I'll suggest very well might not be a good architecture; I'm a relatively fresh (if otherwise experienced) clj dev; I've actually not used rum extensively, only on a basic level)*

In Rum, you can extend components (defined via `rum/defc(s)`) with mixins ([Docs](https://github.com/tonsky/rum#writing-your-own-mixin)). One such mixin is the [reactive mixin (code link)](https://github.com/tonsky/rum/blob/gh-pages/src/rum/core.cljs#L352-L391). You add it to a component and use it [like so (docs link)](https://github.com/tonsky/rum#reactive-components).

Theoretically, it should be possible to define a mixin pretty much analogous to the reactive mixin, but probably a lot simpler. It'd be simpler because something like the helper function `(rum/react <atom>)` (which adds the atom to be watched) wouldn't be necessary, since you already have the atoms the component depends on in your code, and can simply register watches on them on first render. You could even keep using reagent atoms and keep almost everything existing the same ¯\\\_(ツ)_/¯

If I understand correctly, the Stylefy Rum mixin would need to:

1. Add the `style` and/or `class` tags to the element (in probably pretty much the same way `use-styles` works now?)
2. Register a watch on the CSS class atoms to force a re-render on the component when the styles change
3. Remove watches on the now-unused atoms when the styles change

This mixin could be used instead of the (now Reagent-only?) `use-style` fn, like so:

```
(rum/defcs my-styled-component
  < (stylefy.rum/use-styles {:color "red"}) ; normal "use-styles" args
  []
  [:h1 "Hello World"])
```

As far as project organization goes, you mentioned that you didn't like to have "two versions of the project". And I don't think you'd need that. I think there might actually be two options, a "dirtier" and a "cleaner" approach (but maybe there are more, as I said, I'm very inexperienced in clj/cljs).

Dirty: Since Rum mixins are completely data driven, you don't need a dependency on Rum to define one, only on React for the "React component", which you already have in there. So you could simply change the `dom` namespace a bit (so that there can be access to the atoms from somewhere else), add a new `stylefy.rum` namespace and add the mixin there.

Clean: Properly break out the `use-styles` functionality into some kind of interface (something along the lines of the three concerns I mentioned above, those should be pretty much the same between Reagent + Rum, right?). Then you can create a new Reagent and Rum module/library/dependency that people would add to their projects, each containing the respective implementation of the interface. I'd believe a lot of the code could be in shared space. I'm pretty sure I've seen this approach by some library already (for Reagent, Om and Rum), but I can't for the life of me remember what it was.

Those are my thoughts. Theoretically, I'd be interested to add this myself, however I'm really really really not a great clj/cljs dev right now and, like everyone, am busy with other stuff. Maybe if I'm at least confident that the first part of that previous sentence is not the case anymore :D

**Comment by Azzurite** [comment]
Actually, I just noticed that the `impl` namespace and vars inside those namespaces are not actually private (seems like most everything is public in clj?) so I might try to write such a mixin myself tomorrow...

**Comment by Azzurite** [comment]
Also, you wrote in another issue:

>  stylefy depends on Reagent because stylefy stores all its styles in Reagent atoms. This makes sure that every time this atom is changed, components are re-rendered, because they depend on this Reagent atom. If this re-rendering functionality is broken, then all styles are returned to components as inline styles. This works for most styles, but for example media queries cannot be used in inline styles. This is one unfortunate limitation in CSS. To bypass it, stylefy converts styles to CSS classes and media queries asynchronously, and then uses Reagent atoms to re-render components using the converted CSS code.

I have no idea how it exactly works,  but I've been trying to use Stylefy with Rum, and I definitely don't see inline styles in my components. Everything seems to work fine, until I started to use media queries, at which point *sometimes* the component still has inline styles and is not visible.

I could imagine it has something to do with [this](https://github.com/tonsky/rum#performance).

**Comment by Jarzka** [comment]
I have to admit that I have never used Rum, but since it has been the most requested feature for stylefy, it would be time for me to take look at it. Every Clojure(Script) project I have worked during the last 6 years has used Reagent, so there hasn't been much pressure for me to touch Rum. :-)

As you probably noticed: the only reason the library currently depends on Reagent it's the re-render mechanism. In short: styles being used are added to into the DOM asynchronously for speed reasons. Components that are using styles that are currently not yet added into the DOM, are hidden for a short period of time until the DOM is ready. Because of this, components need to watch and see when a specific style is ready, and then re-render. With Reagent, this easy to do with Reagent atom dereference. For Rum, we need something else.

The advice you gave on using stylefy with Rum looks correct, but since I'm not familiar with Rum I cannot say for sure. :) I'll give more comments when I have tested Rum and stylefy together.

> it very well might not be a good architecture

I'm sorry to hear that, I have tried to make the code as easy to understand as I possibly could, even if this domain is not necessarily easy to approach.

> Clean: Properly break out the use-styles functionality into some kind of interface (something along the lines of the three concerns I mentioned above, those should be pretty much the same between Reagent + Rum, right?). Then you can create a new Reagent and Rum module/library/dependency that people would add to their projects, each containing the respective implementation of the interface. I'd believe a lot of the code could be in shared space. I'm pretty sure I've seen this approach by some library already (for Reagent, Om and Rum), but I can't for the life of me remember what it was.

I like this approach of not creating a different stylefy version for every different React implementation, but instead make the library more "modular" to be usable with different React libraries (actually, that's what I already did with dom.cljs when I implemented SSR support). As you said, we could still have a default implementation for the most used libraries (Reagent, Rum...) so it wouldn't make the setup process that much more difficult. 

> I have no idea how it exactly works, but I've been trying to use Stylefy with Rum, and I definitely don't see inline styles in my components. Everything seems to work fine, until I started to use media queries, at which point sometimes the component still has inline styles and is not visible.

I believe Rum does not depend on Reagent atoms and thus does not re-render components when Reagent atoms change. Components using styles containing media queries are simply hidden until the style has been generated and added into the DOM. After that, the re-rendering never happens and thus the component stays visible. However, if you have stylefy's caching functionality turned on (it's on by default), this problem only occurs the first the a specific style is generated. After that, it's loaded from cache

Theoretically stylefy and Rum would work together if we simply render everything synchronously (no need for watcher etc.). However, this does slow down the render process a lot, I don't see it as an option. Caching the generated styles helps, but since HTML5 local storage it not actually meant to be used for caching, I like to keep the caching functionality optional (even if it's turned on by default). IndexedDB would be a better place for caching, but again: I wouldn't like add it as a mandatory requirement for using this library.

**Comment by Jarzka** [comment]
Rum's reactive mixings could probably be used as an easy replacement for Reagent atom derefences. However, if I understand correctly, it requires that every single component should be defined as a reactive component, if you want to use stylefy. I'm not sure if this is a problem?

**Comment by Jarzka** [comment]
There is now a working, yet still pretty much work-in-progress branch for Rum testing:
https://github.com/Jarzka/stylefy/tree/feature/rum

What I did is that I extracted the whole dom.cljs into reagent and rum versions. Both of them are pretty similar, only the deref-implementation is a bit different. I'm not sure if this will be the final way to do this, but theoretically if we modularize the whole dom.cljs, we can implement support for any React/non-React UI framework we want. We could even have a generic DOM module which uses synchronous DOM manipulation for smaller applications.

As I said, there is still a lot of work to be done in order for this to be production-ready:
- stylefy/font-face, stylefy/keyframes etc. do not currently work if you call them before stylefy has been initialised. That is because the current DOM module needs to be initialised before any CSS can be generated. 
- ReagentDOM and RumDOM are now located in examples, they should be moved into their own modules than can be used individually in proejcts.

Still, I think it looks promising.

**Comment by Azzurite** [comment]
> it very well might not be a good architecture

I was talking about what I was about to suggest, not this library :D as I said, I'm not a good clj dev so I was basically just saying that what I'm about to suggest may be bad :D For what it's worth, when I looked through stylefy code, I could at least quickly find the places where the reagent integration / rerendering was happening (though it took a little to actually get that calling the `style-in-dom?` method is what actually sets up the re-rendering dependency).

>Maybe similar modularization like in reitit 

This was exactly the library I was talking about in the main post that I couldn't remember!

> every single component should be defined as a reactive component, if you want to use stylefy. I'm not sure if this is a problem. Probably not

I'm pretty sure this wouldn't be a problem. Anyone can just write themselves a little helper macro that does that automatically, if they wanted to.

> There is now a working, yet pretty much work-in-progress branch for Rum testing

Are you some kind of superhuman? :) I will check this out and see how it works!

**Comment by Jarzka** [comment]
I separated the different DOM namespaces into modules, those can be found in the `modules` folder now, just like in reitit. I didn't try to install the modules manually to include them as dependencies (yet), but those are defined the example projects' source paths.

> Are you some kind of superhuman? :) I will check this out and see how it works!

I just happen to have some freetime during this weekend so implemented Rum support is a nice challenge to spend time with. :P

**Comment by Azzurite** [comment]
So yeah I've not actually figured out how to run the examples namespace or even build your project, but I've added a `deps.edn` like this:

```
{:paths ["src/clj" "src/cljc" "src/cljs" "examples-rum/src/cljs"]
 :deps {org.clojure/clojure {:mvn/version "1.9.0"}
        org.clojure/clojurescript {:mvn/version "1.9.946"}
        prismatic/dommy {:mvn/version "1.1.0"}
        hiccup {:mvn/version "1.0.5"}
        reagent {:mvn/version "0.7.0"}
        garden {:mvn/version "1.3.6"}
        org.clojure/core.async {:mvn/version "0.3.443"}}
  :aliases {:cljs {:extra-deps {thheller/shadow-cljs {:mvn/version "2.11.18"}}}
            :dev {:extra-deps {clj-chrome-devtools {:mvn/version "20180310"}}}}}
```

and then added it as a local dependency (`stylefy/stylefy {:local/root "../stylefy"}`) in my project. It seems to work in a basic way but there's definitely some issues. Compilation errors + warnings/errors in the console and stuff. 

Do you want any of my feedback? Do you want it here? Should I wait until you're farther along?

**Comment by Jarzka** [comment]
I think right now the easiest way to test the result is to clone the repo, switch to `feature/rum` branch and run the example project. You should be example to go to `examples/rum/`, run `shadow-cljs watch frontend` and point your browser to `localhost:8000` to see the result. That's how I'm testing the implementation until the work progresses a bit further.

**Comment by Jarzka** [comment]
Another problem solved: stylefy/font-face, stylefy/keyframes etc. can now be called before initialising stylefy, just like before. The values are saved temporarily until the real DOM handler has been attached.

**Comment by Jarzka** [comment]
Here is how to test the modules in your own project: clone the repo, switch to `feature/rum` branch and run `lein install`. Then go to modules/stylefy-rum and run `lein install` there as well. These commands will install the libraries locally (just like you would download them from clojars). After that, you should be able to include the modules in your own project:

```
[stylefy "3.0-SNAPSHOT"]
[stylefy-rum "3.0-SNAPSHOT"]
```

**Comment by Azzurite** [comment]
So I just plugged it in there. I'm doing SSR with Rum & Stylefy, then hydrating it on the client.

The first thing I see is this:

```
react-dom.development.js:68 Warning: Cannot update during an existing state transition (such as within `render`). Render methods should be a pure function of props and state.
    at azzurite.web.content.homepage/page (http://192.168.178.53:8080/js/cljs-runtime/rum.core.js:119:185)
    at azzurite.web.content.homepage/homepage (http://192.168.178.53:8080/js/cljs-runtime/rum.core.js:119:185)
printWarning @ react-dom.development.js:68
error$jscomp$0 @ react-dom.development.js:44
scheduleUpdateOnFiber @ react-dom.development.js:24012
enqueueForceUpdate @ react-dom.development.js:12505
Component.forceUpdate @ react.development.js:385
eval @ core.cljs:377
eval @ core.cljs:4450
cljs$core$reset_BANG_ @ core.cljs:4491
eval @ core.cljs:4525
stylefy$rum$dom$save_style @ dom.cljs:102
eval @ dom.cljs:154
stylefy$impl$dom$save_style @ dom.cljs:37
eval @ core.cljc:68
stylefy$impl$styles$create_style_BANG_ @ styles.cljc:23
stylefy$impl$styles$use_style_BANG_ @ styles.cljc:122
eval @ core.cljc:68
eval @ core.cljc:64
eval @ homepage.cljc:14
eval @ core.cljs:3918
eval @ core.cljs:3913
eval @ core.cljs:3948
render @ core.cljs:201
eval @ core.cljs:366
eval @ core.cljs:105
eval @ core.cljs:105
finishClassComponent @ react-dom.development.js:17486
updateClassComponent @ react-dom.development.js:17436
beginWork @ react-dom.development.js:19074
beginWork$1 @ react-dom.development.js:23941
performUnitOfWork @ react-dom.development.js:22780
workLoopSync @ react-dom.development.js:22708
renderRootSync @ react-dom.development.js:22671
performSyncWorkOnRoot @ react-dom.development.js:22294
scheduleUpdateOnFiber @ react-dom.development.js:21882
updateContainer @ react-dom.development.js:25483
eval @ react-dom.development.js:26022
unbatchedUpdates @ react-dom.development.js:22432
legacyRenderSubtreeIntoContainer @ react-dom.development.js:26021
exports.hydrate @ react-dom.development.js:26087
rum$core$hydrate @ core.cljs:229
azzurite$web$frontend$core$render @ core.cljs:9
azzurite$web$frontend$core$init @ core.cljs:14
eval @ shadow.module.all.append.js:4
goog.globalEval @ site.js:577
env.evalLoad @ site.js:1670
(anonymous) @ site.js:1883
```

This seems like something stylefy is doing.

Then we have 

```
react-dom.development.js:68 Warning: Prop `style` did not match. Server: "null" Client: "color:#FFC975"
    at span
    at azzurite.web.content.homepage/accent-text (http://192.168.178.53:8080/js/cljs-runtime/rum.core.js:119:185)
    at h1
    at div
    at div
    at azzurite.web.content.homepage/container (http://192.168.178.53:8080/js/cljs-runtime/rum.core.js:119:185)
    at div
    at azzurite.web.content.homepage/welcome (http://192.168.178.53:8080/js/cljs-runtime/rum.core.js:119:185)
    at div
    at azzurite.web.content.homepage/page (http://192.168.178.53:8080/js/cljs-runtime/rum.core.js:119:185)
    at azzurite.web.content.homepage/homepage (http://192.168.178.53:8080/js/cljs-runtime/rum.core.js:119:185)
printWarning @ react-dom.development.js:68
error$jscomp$0 @ react-dom.development.js:44
warnForPropDifference @ react-dom.development.js:8825
completeWork @ react-dom.development.js:9596
completeUnitOfWork @ react-dom.development.js:22813
performUnitOfWork @ react-dom.development.js:22788
workLoopSync @ react-dom.development.js:22708
renderRootSync @ react-dom.development.js:22671
performSyncWorkOnRoot @ react-dom.development.js:22294
scheduleUpdateOnFiber @ react-dom.development.js:21882
updateContainer @ react-dom.development.js:25483
eval @ react-dom.development.js:26022
unbatchedUpdates @ react-dom.development.js:22432
legacyRenderSubtreeIntoContainer @ react-dom.development.js:26021
exports.hydrate @ react-dom.development.js:26087
rum$core$hydrate @ core.cljs:229
azzurite$web$frontend$core$render @ core.cljs:9
azzurite$web$frontend$core$init @ core.cljs:14
eval @ shadow.module.all.append.js:4
goog.globalEval @ site.js:577
env.evalLoad @ site.js:1670
(anonymous) @ site.js:1883
```

Not sure if this is something I'm doing wrong, but the color there worked perfectly fine before without a mismatch between server/client.

**Comment by Azzurite** [comment]
I don't know if you want to see it, but the code for that is here: https://gitlab.com/azzurite/web

**Comment by Azzurite** [comment]
Also a new thing:

```
react-dom.development.js:68 Warning: Prop `className` did not match. Server: "_stylefy_2043097725" Client: "_stylefy_1190614352"
    at span
    at p
    at div
    at azzurite.web.content.homepage/container (http://192.168.178.53:8080/js/cljs-runtime/rum.core.js:119:185)
    at div
    at azzurite.web.content.homepage/page (http://192.168.178.53:8080/js/cljs-runtime/rum.core.js:119:185)
    at azzurite.web.content.homepage/homepage (http://192.168.178.53:8080/js/cljs-runtime/rum.core.js:119:185)
printWarning	@	react-dom.development.js:68
error$jscomp$0	@	react-dom.development.js:44
warnForPropDifference	@	react-dom.development.js:8825
completeWork	@	react-dom.development.js:9646
completeUnitOfWork	@	react-dom.development.js:22813
performUnitOfWork	@	react-dom.development.js:22788
workLoopSync	@	react-dom.development.js:22708
renderRootSync	@	react-dom.development.js:22671
performSyncWorkOnRoot	@	react-dom.development.js:22294
scheduleUpdateOnFiber	@	react-dom.development.js:21882
updateContainer	@	react-dom.development.js:25483
eval	@	react-dom.development.js:26022
unbatchedUpdates	@	react-dom.development.js:22432
legacyRenderSubtreeIntoContainer	@	react-dom.development.js:26021
exports.hydrate	@	react-dom.development.js:26087
rum$core$hydrate	@	core.cljs:229
azzurite$web$frontend$core$render	@	core.cljs:9
azzurite$web$frontend$core$init	@	core.cljs:14
eval	@	shadow.module.all.append.js:4
goog.globalEval	@	site.js:577
env.evalLoad	@	site.js:1670
(anonymous)	@	site.js:1883
```

**Comment by Azzurite** [comment]
> I would appreciate to get some feedback how Rum and stylefy work together

But yeah, I'm not doing anything particularly special so far. don't even have any dynamic styles or anything more advanced. I'll just keep using this version and let you know about anything I notice.

**Comment by Jarzka** [comment]
I have to test run your project at some point to get a better idea what is failing there. It will probably take a few days until I have time to work with this feature again.

The good news is that I was able to run my personal website with stylefy SSR along with the brand new stylefy-reagent module. So at least something is working already. :-)

**Comment by Azzurite** [comment]
Those are all warnings I posted, it is working already. Guess the console log export did not add "warning:" in front.

So it is working for Rum in total. I'm using it to continue developing right now. Just maybe not 100%.

**Comment by Azzurite** [comment]
Also just so you know, since you defined the projects as `stylefy-rum` and `stylefy-reagent`, the generated maven artifact has the group name `stylefy-rum` as well. In clojure tools `deps.edn`, it is deprecated to not use the group name when specifying dependencies, so currently my `deps.edn` looks like this:

```
stylefy/stylefy {:mvn/version "3.0-SNAPSHOT"
stylefy-rum/stylefy-rum {:mvn/version "3.0-SNAPSHOT"}
```

Maybe it'd be a good idea to make the project name `stylefy/rum` and `stylefy/reagent` instead or something like that.

**Comment by Jarzka** [comment]
> Those are all warnings I posted, it is working already.

Oh okay, I thought the warnings were causing more problems. Good to hear it's working. :) Still, it would be good idea to check if there are issues behind those warnings.

> Maybe it'd be a good idea to make the project name stylefy/rum and stylefy/reagent instead or something like that.

Thanks for the tip!

**Comment by Jarzka** [comment]
A few quick updates:
- Module names updated
- Namespaced keywords replaced with unnamespaced when caching CSS. This makes different DOM module implementations easier to work with, and it also reduces the size of the cache approximately 7%. Unsupported cache versions are cleared automatically, so no actions are required from the users of the library.
- Slightly recuded the amount of duplicated code in different DOM modules

**Comment by Jarzka** [comment]
I started porting the included Reagent examples to Rum to get a better understanding of the supported functionality. I have also used the example projects as a manual testing tool while updating the library.

I'm already seeing the same warning message you got:
``` Cannot update during an existing state transition```
So now it's easy for me to repeat it and investigate it a bit further.

I also noticed there is currently some compile problem with rum and shadow-cljs:
https://github.com/tonsky/rum/issues/233

Thus, I'm using lein-cljsbuild for now so that I can continue working with this.

**Comment by Azzurite** [comment]
Still really appreciating the updates :)

> I also noticed there is currently some compile problem with rum and shadow-cljs

you can use `0.12.3` for now until that is fixed.

**Comment by Jarzka** [comment]
I have now ported all the Reagent examples to Rum. Everything seems to work as expected, so I think we will be ready to release a public beta version soon.

> Cannot update during an existing state transition

This warning is given because style updates its internal "style-in-dom" state during render. This is a valid warning, because normally do not want to reset states during rendering, but would consider stylefy as an exception. I tried updating the internal state of the library asynchronously, outside of any rendering functions. This caused the warning message to go away, but it also caused all components render themselves way too many times because they were all reacting to the changes in the "style-in-dom" atom. I'm not sure if we have to accept the warning since I cannot think an easy fix for this.

**Comment by Jarzka** [comment]
The next beta version of stylefy, along with the new modules, is now publicly available:

```
[stylefy "3.0.0-beta1"]
[stylefy/reagent "3.0.0-beta1"] 
[stylefy/rum "3.0.0-beta1"] 
[stylefy/generic-dom "3.0.0-beta1"]
```

**Comment by Azzurite** [comment]
> probably related to how Rum handles server-side components

I'd guess they're not, because it only happens when you do something stupid like `(use-style {:font-size 8})` instead of `(use-style {:font-size "8px"})`. I remember reading something about react automatically converting a number value to "<number>px". Whatever it is, I guess it's handling styles with numbers as values differently between clj and cljs. Could still be related to rum, I haven't tested that with reagent yet. But yeah since `8` -> `"8px"` fixes it, it doesn't matter to me.

**Comment by Jarzka** [comment]
> I'd guess they're not, because it only happens when you do something stupid like (use-style {:font-size 8}) instead of (use-style {:font-size "8px"}). I remember reading something about react automatically converting a number value to "px".

I see, this sounds like a possible explanation.

**Comment by Jarzka** [comment]
`[stylefy/rum "3.0.0-beta2"]` released. Reduces amount of "useless" rendering.

**Comment by Jarzka** [comment]
Now the optimisation works so here is `[stylefy/rum "3.0.0-beta4"]`. Now components re-render only when their own styles are added into DOM, not every time when some style is added into DOM. This optimisation is already in the Reagent module, but it required a bit of learning Rum before I was able implement it for the Rum module.

**Comment by Jarzka** [comment]
`[stylefy/rum "3.0.0-beta5"]` is out. It uses `react-dom/unstable_batchedUpdates` to update all "style-in-dom" states, so that components render only after all states have been updated. Not it works the same way as with Reagent.

However, since `batchedUpdates` is marked unstable and we don't know if it's going to removed or not, I future-proofet the solution in a way that if this `unstable_batchedUpdates` does not exists in `react-dom`, we simply do the update without it.

At this point, the `stylefy-rum` contains all the optimisations and works pretty much the same way as `style-reagent`.

**Comment by Azzurite** [comment]
Very nice!

I now only have a "real" `Warning: Prop `className` did not match. Server: "_stylefy_-1339687936" Client: "_stylefy_1008907027"` problem for server-side rendering. The style is `only-on-desktop`:
```
(def max-mobile-device-width (px 768))
(def min-desktop-breakpoint (g/+ 1 max-mobile-device-width))
(def only-on-desktop {:display "none"
                      ::stylefy/media {{:min-width min-desktop-breakpoint}
                                       {:display "unset"}}})
```

I can fix it the same way as before, if I change `min-desktop-breakpoint` from a garden `px` unit to a string `"769px"`, then the class name matches again. However, this time, garden units are meant to be officially supported, so I don't really want to change to string again.

Also I get "WARNING: class already refers to: #'clojure.core/class in namespace: stylefy.core, being replaced by: #'stylefy.core/class" when starting up my REPL, would it maybe make sense to add `(:refer-clojure :exclude [class])` to your `(ns)` in that file?

**Comment by Jarzka** [comment]
> Also I get "WARNING: class already refers to: #'clojure.core/class in namespace: stylefy.core, being replaced by: #'stylefy.core/class" when starting up my HTTP server in the (Clojure) REPL, would it maybe make sense to add (:refer-clojure :exclude [class]) to your (ns) in that file?

This warning has been around for a long time, but I haven't really paid much attention to it. Good to know there is a way to get rid of it by excluding `clojure.core/class`! I have to try it. :)

> I can fix it the same way as before, if I change min-desktop-breakpoint from a garden px unit to a string "769px", then the class name matches again. However, this time, garden units are meant to be officially supported, so I don't really want to change to string again.

This sounds strange because I believe using Garden units should end up as strings in CSS and should not cause this kind of warnings. I have to take a closer look at this... hopefully in the coming days.

**Comment by Jarzka** [comment]
`clojure.core/class` conflict is now fixed in the code. The warning will go away with the next release of stylefy core.

> Warning: Prop `className` did not match. Server: "_stylefy_-1339687936" Client: "_stylefy_1008907027"

I just tested that the hashing result should be the same no matter if you use Garden units or write the unit manually by hand:

```cljs
stylefy.impl.hashing=> (hash-style {:width "100px"})
"_stylefy_955709200"
stylefy.impl.hashing=> (hash-style {:width (garden.units/px 100)})
"_stylefy_955709200"
```

Can I test your project in order to investigate this a bit more?

**Comment by Azzurite** [comment]
> Can I test your project in order to investigate this a bit more?

Of course!

https://gitlab.com/azzurite/web

You need to have clojure tools & npm installed. Simply clone & just follow the 3 steps in the readme.

**Comment by Jarzka** [comment]
Actually, I think I figured it out. :D Garden units are not converted to CSS before hashing if they are nested deep inside a style map:

````cljs
(hash-style {:width "100px"
             :stylefy.core/media {{:max-width "500px"}
                                  {:width "100px"}}})

"_stylefy_-1822457317"

(hash-style {:width "100px"
             :stylefy.core/media {{:max-width "500px"}
                                  {:width (garden.units/px 100)}}})
"_stylefy_490980497"
```

So to fix this, the hashing function should also check nested maps / vectors for convertable Garden units.

**Comment by Jarzka** [comment]
I pushed a change into `feature/rum` branch which will check Garden values recursively during hashing. I think this fixes it, at least my unit test passes.

Could you please test it by installing the stylefy core module locally from the `feature/rum` branch?

**Comment by Azzurite** [comment]
> Could you please test it by installing the stylefy core module locally from the feature/rum branch?

I checked out f5858d3c0ea5857640ccf4dd939708d7425e2bcf adjusted the `project.clj` to `3.0.0-beta2` did a `lein install`, then went to my project and specified `3.0.0-beta2` as `stylefy/stylefy` dependency and tried it with that version.

The "className did not match" error still appears at [the same place](https://gitlab.com/azzurite/web/-/blob/master/src/azzurite/web/content/homepage.cljc#L192) with the above mentioned style. Didn't seem to help.

**Comment by Jarzka** [comment]
I think the easiest way to check if you have a correct version, is to check the project.clj of style core module and change `(defproject stylefy "3.0.0-beta1"` to something like `(defproject stylefy "3.0.0-SNAPSHOT20210406"`. Then do `lein install` and make sure your project also depends on `stylefy/style "3.0.0-SNAPSHOT20210406"`

**Comment by Azzurite** [comment]
Yeah that's what I did basically just with `3.0.0-beta2`.

I do have `caching` on though. If I clear the cache in the browser, then I get the `Warning: Prop `style` did not match. Server: "null" Client: "color:#ffc975"` at [this place](https://gitlab.com/azzurite/web/-/blob/master/src/azzurite/web/content/homepage.cljc#L189) ([`accent-color`](https://gitlab.com/azzurite/web/-/blob/master/src/azzurite/web/content/homepage.cljc#L24)) instead. This is only on the first page open after cache clearing. When I then press `F5` again, the `className did not match` warning appeared again.

**Comment by Jarzka** [comment]
I was able to compile your project and get Dashboard running correctly. How I'm supposed to see the warning message?

**Comment by Azzurite** [comment]
Oh, the browser console? In the devtools. Should open with F12

**Comment by Jarzka** [comment]
![image](https://user-images.githubusercontent.com/5528061/113735896-544e9980-9705-11eb-9b41-27a6ee72a750.png)

I cannot see anything, not even stylefy classes anywhere :)

**Comment by Azzurite** [comment]
Oh, you need to go to `localhost:8080`.

**Comment by Jarzka** [comment]
I just saw this on the console:
`shadow-cljs - server version: 2.11.18 running at http://localhost:9630`
and clicked the link and wondered what I'm supposed to do. :P 

Anyway, now I see the warning. I'll get back when I taken a closer look.

**Comment by Jarzka** [comment]
I think it might be related to the fact that you are using Garden value in a map, which itself is also a key. Maps that are keys are not checked for Garden values. Very interesting finding! I'll try to fix this too. :D

**Comment by Jarzka** [comment]
> Warning: Prop `className` did not match. Server: "_stylefy_-1339687936" Client: "_stylefy_1008907027"

I pushed a fix which will get rid of this warning as the hash should now be equal on both frontend and backend.

> Warning: Prop `style` did not match. Server: "null" Client: "color:#ffc975"

This still needs a bit of investigation.

**Comment by Jarzka** [comment]
Warning: Prop style did not match. Server: "null" Client: "color:#ffc975"

This happens because `use-style` works a bit differently on frontend and backend. On the backend, the style is converted to CSS, inserted into the page's `head`, and the generated class name is returned. On the frontend, inline style is returned until the style has been added into DOM.

I believe this is also a warning that is difficult to get rid of. It's not even a warning in our case because the rendering works the way it's designed. I think the only way to get rid of it is to use the generic `stylefy/generic-dom` module, which does only synchronous DOM updates and does not suffer from warnings like this. But once again, the wide support comes with performance cost.

**Comment by Azzurite** [comment]
> I pushed a fix which will get rid of this warning as the hash should now be equal on both frontend and backend

Yep, that fixed it. 👍 Looking really great so far, I'm impressed!

> It's not even a warning in our case

Yeah, makes sense, the only important thing is just that hydrating works and all functionality afterwards is the same, which it looks like to me.

> I believe this is also a warning that is difficult to get rid of

I would imagine the "ideal" way would be to somehow detect that the rendering is happening while `hydrate`-ing and just not bother with inline styles in that case.

**Comment by Jarzka** [comment]
Good to hear that the fix works. I think we are coming close to release the version `3.0.0`, without "beta" :)

**Comment by Jarzka** [comment]
3.0.0 is now officially released.

**Comment by Jarzka** [comment]
Closing this as done. If issues arise, let's discuss them in a new thread.

**Comment by Azzurite** [comment]
Oh yeah, of course! Sorry, forgot about this. I've been using it for a while and it works really well! Really nice job :)
